<a href="https://colab.research.google.com/github/dawit-melka/ICog-Labs/blob/main/semantic_search.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Text Embedding Models Analysis**

In [ ]:
!git clone https://github.com/dawit-melka/ICog-Labs.git

In [ ]:
%cd ICog-Labs

In [ ]:
%cd Semantic Search

In [ ]:
!pip install fuzzywuzzy
!pip install sentence_transformers
!pip install voyageai
!python -m pip install -U angle-emb

In [ ]:
from sentence_transformers import SentenceTransformer, util
import os
import nltk
import time
import matplotlib.pyplot as plt
import numpy as np
from fuzzywuzzy import fuzz

import voyageai
from voyageai import get_embeddings, get_embedding

from data.questions_and_answers import questions, answers
nltk.download('punkt')  # Uncomment if not already downloaded


In [23]:
def load_input_data(file_path):
    """
    Load input data from a file.

    Parameters:
    - file_path (str): Path to the input file.

    Returns:
    - str: Contents of the file.
    """
    with open(file_path, 'r', encoding='utf-8') as f:
        data = f.read()
    return data

In [24]:
def convert_to_sentences(text):
    """
    Convert a text into a list of sentences.

    Parameters:
    - text (str): Input text.

    Returns:
    - list: List of sentences.
    """
    return nltk.sent_tokenize(text)

In [25]:
def download_models(model_names):
    """
    Download sentence transformer models.

    Parameters:
    - model_names (list): List of model names to download.
    """
    for model_name in model_names:
        model = SentenceTransformer(model_name)


In [26]:
def compute_similarity_scores(query_emb, doc_emb):
    """
    Compute similarity scores between a query and document embeddings.

    Parameters:
    - query_emb (numpy array): Embedding of the query.
    - doc_emb (numpy array): Embeddings of the documents.

    Returns:
    - list: Similarity scores.
    """
    return util.dot_score(query_emb, doc_emb)[0].cpu().tolist()

In [27]:
def check_similarity(sentence1, sentence2):
    similarity_ratio = fuzz.ratio(sentence1, sentence2)

    threshold = 70

    if similarity_ratio >= threshold:
        return True
    else:
        return False


In [28]:
def save_model_stats_to_txt(model_stats, file_path):
    with open(file_path, 'a') as file:
        for model_name, stats in model_stats.items():
            file.write(f"Model: {model_name}\n")
            file.write(f"Time Elapsed: {stats['time elapsed']} seconds\n")
            file.write(f"Total Error: {stats['total error']}\n")
            file.write(f"Total Found: {stats['total found']}\n")
            file.write(f"First Results: {stats['first results']}\n")
            file.write(f"Second Results: {stats['second results']}\n")
            file.write(f"Third Results: {stats['third results']}\n")
            file.write("\n")

In [29]:
# Load input data
INPUT_FILE_PATH = '/content/ICog-Labs/Semantic Search/data/to_kill_a_mocking_bird.txt'
data = load_input_data(INPUT_FILE_PATH)

# Convert the book into a list of sentences
book_sentences = convert_to_sentences(data)

In [30]:
def run_semantic_search(model_name, model):
    model_stats = {}

    start_time = time.time()
    print(model_name, "model running...")


    # Encode query and documents
    if model_name == "embed-english-v3.0":
      doc_emb = model.embed(book_sentences, input_type="search_document", model=model_name).embeddings
    elif model != None:
      doc_emb = model.encode(book_sentences)
    else:
      doc_emb = [get_embedding(s) for s in book_sentences]
    error = 0
    count_first_result = 0
    count_second_result = 0
    count_third_result = 0
    count_top_10 = 0

    # Implementation of the semantic search logic...
    for i, query in enumerate(questions):
        if model_name == "embed-english-v3.0":
          query_emb = model.embed([query], input_type="search_query", model=model_name).embeddings
        elif model != None:
          query_emb = model.encode(query)
        else:
          query_emb = get_embedding(query)

        # Compute dot score between query and all document embeddings
        scores = compute_similarity_scores(query_emb, doc_emb)

        # Combine docs & scores
        doc_score_pairs = list(zip(book_sentences, scores))

        # Sort by decreasing score
        doc_score_pairs = sorted(doc_score_pairs, key=lambda x: x[1], reverse=True)

        # Get top-k results
        top_k = 10
        found = False

        answer_sentences = convert_to_sentences(answers[i])
        for j in range(1, top_k + 1):
            for a in answer_sentences:
              isSimmilar= check_similarity(doc_score_pairs[j][0], a)
              if isSimmilar:
                  error += (j - 1)
                  found = True
                  count_first_result += 1 if j == 1 else 0
                  count_second_result += 1 if j == 2 else 0
                  count_third_result += 1 if j == 3 else 0
                  count_top_10 += 1
                  break
            if found:
              break

        if not found:
            error += 15
        # print(i + 1, "/", len(questions), " error: ", error, "count found: ", count_top_10)

    end_time = time.time()
    model_stats[model_name] = {"time elapsed": end_time - start_time,
                                "total error": error,
                                "total found": count_top_10,
                                "first results": count_first_result,
                                "second results": count_second_result,
                                "third results": count_third_result}
    print(model_name," model status \n", model_stats[model_name])

    stats_file_path = '/content/ICog-Labs/Semantic Search/data/model_stats.txt'
    save_model_stats_to_txt(model_stats, stats_file_path)

    return model_stats

In [ ]:
def run_sbert_models(INPUT_FILE_PATH):
    # Define constants
    # INPUT_FILE_PATH = os.path.join(os.path.dirname(__file__), '.\\data\\to_kill_a_mocking_bird.txt')
    MODELS = ['multi-qa-mpnet-base-dot-v1',
              'all-mpnet-base-v2',
              'gtr-t5-xl',
              'all-roberta-large-v1',
              'multi-qa-distilbert-cos-v1',
              'all-distilroberta-v1',
              'msmarco-bert-base-dot-v5',
              'multi-qa-MiniLM-L6-cos-v1',
              'msmarco-distilbert-dot-v5',
              'all-MiniLM-L12-v2',
              ]

    # Load input data
    data = load_input_data(INPUT_FILE_PATH)

    # Convert the book into a list of sentences
    book_sentences = convert_to_sentences(data)

    # Download all models
    download_models(MODELS)

    model_stats = {}

    for model_name in MODELS:
        start_time = time.time()
        print(model_name, "model running...")

        # Load the model
        model = SentenceTransformer(model_name)

        # Encode query and documents
        doc_emb = model.encode(book_sentences, convert_to_tensor=True)

        error = 0
        count_first_result = 0
        count_second_result = 0
        count_third_result = 0
        count_top_10 = 0

        # Implementation of the semantic search logic...
        for i, query in enumerate(questions):
            query_emb = model.encode(query, convert_to_tensor=True)

            # Compute dot score between query and all document embeddings
            scores = compute_similarity_scores(query_emb, doc_emb)

            # Combine docs & scores
            doc_score_pairs = list(zip(book_sentences, scores))

            # Sort by decreasing score
            doc_score_pairs = sorted(doc_score_pairs, key=lambda x: x[1], reverse=True)

            # Get top-k results
            top_k = 10
            found = False
            print("Query: ", query)
            print("Answer: ", answers[i])
            answer_sentences = convert_to_sentences(answers[i])
            for j in range(1, top_k + 1):
                for a in answer_sentences:
                  isSimmilar= check_similarity(doc_score_pairs[j][0], a)
                  if isSimmilar:
                      error += (j - 1)
                      found = True
                      count_first_result += 1 if j == 1 else 0
                      count_second_result += 1 if j == 2 else 0
                      count_third_result += 1 if j == 3 else 0
                      count_top_10 += 1
                      break
                if found:
                  break

            if not found:
                error += 15
            # print(i + 1, "/", len(questions), " error: ", error, "count found: ", count_top_10)

        end_time = time.time()
        model_stats[model_name] = {"time elapsed": end_time - start_time,
                                   "total error": error,
                                   "total found": count_top_10,
                                   "first results": count_first_result,
                                   "second results": count_second_result,
                                   "third results": count_third_result}
        print(model_name, " model status \n", model_stats[model_name])

    print(model_stats)
    stats_file_path = '/content/ICog-Labs/Semantic Search/data/model_stats.txt'
    save_model_stats_to_txt(model_stats, stats_file_path)




In [ ]:
INPUT_FILE_PATH = '/content/ICog-Labs/Semantic Search/data/to_kill_a_mocking_bird.txt'
run_sbert_models(INPUT_FILE_PATH)

In [ ]:
from angle_emb import AnglE

model = AnglE.from_pretrained('WhereIsAI/UAE-Large-V1', pooling_strategy='cls').cuda()
run_semantic_search("UAE-Large-v1", model)

In [ ]:
voyageai.api_key = "VOYAGEAI API KEY HERE"

run_semantic_search("voyage-lite-01-instruct", None)

In [ ]:
!pip install llmx
!pip install -U cohere

In [ ]:
import cohere
cohere_key = "COHERE API KEY HERE"
model = cohere.Client(cohere_key)

In [ ]:
run_semantic_search("embed-english-v3.0", model)

In [ ]:
model = SentenceTransformer('BAAI/bge-large-zh-v1.5')
run_semantic_search("BAAI/bge-large-zh-v1.5", model)

In [ ]:
model = SentenceTransformer('llmrails/ember-v1')
run_semantic_search("llmrails/ember-v1", model)

In [ ]:
model = SentenceTransformer("jamesgpt1/sf_model_e5")
run_semantic_search("jamesgpt1/sf_model_e5", model)

In [ ]:
model = SentenceTransformer("thenlper/gte-large")
run_semantic_search("thenlper/gte-large", model)

In [ ]:
model = SentenceTransformer("thenlper/gte-large")
run_semantic_search("thenlper/gte-large", model)

In [ ]:
model = SentenceTransformer("infgrad/stella-base-en-v2")
run_semantic_search("infgrad/stella-base-en-v2", model)

In [ ]:
!pip install InstructorEmbedding

In [ ]:
from InstructorEmbedding import INSTRUCTOR
model = INSTRUCTOR('hkunlp/instructor-xl')
run_semantic_search("hkunlp/instructor-xl", model)

In [ ]:
model = INSTRUCTOR('hkunlp/instructor-large')
run_semantic_search("hkunlp/instructor-large", model)

In [ ]:
model = INSTRUCTOR('intfloat/e5-large-v2')
run_semantic_search("intfloat/e5-large-v2", model)

In [ ]:
!pip install transformers
from transformers import AutoModel

In [ ]:
model = AutoModel.from_pretrained('jinaai/jina-embeddings-v2-base-en', trust_remote_code=True)
run_semantic_search("jinaai/jina-embeddings-v2-base-en", model)

In [ ]:
model = SentenceTransformer("sentence-t5-xl")
run_semantic_search("sentence-t5-xl", model)

In [31]:
model = SentenceTransformer("BAAI/bge-large-en-v1.5")
run_semantic_search("BAAI/bge-large-en-v1.5", model)

.gitattributes:   0%|          | 0.00/1.52k [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/191 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/90.3k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/779 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.34G [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.34G [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/52.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/711k [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/366 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

BAAI/bge-large-en-v1.5 model running...
BAAI/bge-large-en-v1.5  model status 
 {'time elapsed': 353.5285394191742, 'total error': 235, 'total found': 7, 'first results': 1, 'second results': 1, 'third results': 1}


{'BAAI/bge-large-en-v1.5': {'time elapsed': 353.5285394191742,
  'total error': 235,
  'total found': 7,
  'first results': 1,
  'second results': 1,
  'third results': 1}}

  # Embeding Models Analysis Result
  
  
  | Model                               | Time Elapsed (s) | Total Error | Total Found (21) | First Results | Second Results | Third Results | Dimension     | Size (MB)     |
  |-------------------------------------|------------------|-------------|------------------|---------------|----------------|---------------|---------------|---------------|
  | multi-qa-mpnet-base-dot-v1          | 87.77            | 117         | 16               | 6             | 2              | 2             | 768           | 420           |
  | msmarco-bert-base-dot-v5            | 96.15            | 135         | 14               | 4             | 3              | 3             | 768           | 420           |
  | hkunlp/instructor-xl                | 1595.49          | 158         | 14               | 2             | 0              | 3             | 768           | 4960          |
  | voyage-lite-01-instruct             | 355.76           | 166         | 12               | 2             | 5              | 0             | 1024          | -             |
  | thenlper/gte-large                  | 340.72           | 168         | 12               | 4             | 2              | 1             | 1024          | 670           |
  | multi-qa-distilbert-cos-v1          | 46.22            | 172         | 12               | 3             | 2              | 1             | 768           | 250           |
  | msmarco-distilbert-dot-v5           | 45.93            | 178         | 11               | 2             | 2              | 3             | 768           | 250           |
  | multi-qa-MiniLM-L6-cos-v1           | 14.61            | 183         | 11               | 4             | 0              | 0             | 384           | 80            |
  | hkunlp/instructor-large             | 395.49           | 188         | 11               | 1             | 2              | 2             | 768           | 1340          |
  | gtr-t5-xl                           | 1312.33          | 188         | 10               | 3             | 2              | 0             | 768           | 2370          |
  | all-mpnet-base-v2                   | 89.32            | 191         | 10               | 4             | 1              | 1             | 768           | 420           |
  | all-MiniLM-L12-v2                   | 27.07            | 203         | 10               | 2             | 1              | 0             | 384           | 120           |
  | cohere-embed-english-v3.0           | 10.09            | 194         | 9                | 3             | 2              | 2             | 1024          | -             |
  | all-distilroberta-v1                | 70.17            | 206         | 9                | 0             | 4              | 2             | 768           | 290           |
  | sentence-t5-xl                      | 1517.48          | 209         | 9                | 2             | 2              | 1             | 768           | 2370          |
  | jamesgpt1/sf_model_e5               | 341.54           | 211         | 9                | 3             | 0              | 1             | 1024          | 1340          |
  | UAE-Large-v1                        | 2043.65          | 214         | 8                | 3             | 1              | 2             | 1024          | 1340          |
  | llmrails/ember-v1                   | 354.12           | 216         | 8                | 1             | 3              | 0             | 1024          | 1340          |
  | all-roberta-large-v1                | 485.51           | 234         | 7                | 1             | 1              | 1             | 1024          | 1360          |
  | BAAI/bge-large-en-v1.5              | 353.52           | 235         | 7                | 1             | 1              | 1             | 1024          | 1340          |
  | infgrad/stella-base-en-v2           | 96.30            | 241         | 7                | 1             | 0              | 2             | 768           | 220           |
  | jinaai/jina-embeddings-v2-base-en   | 132.16           | 246         | 6                | 1             | 1              | 0             | 768           | 270           |
  | intfloat/e5-large-v2                | 342.50           | 256         | 5                | 1             | 0              | 1             | 1024          | 1340          |


